src/libstd/collections/hash/table.rs

```diff
@@ -742,7 +742,9 @@ impl<K, V> RawTable<K, V> {
     ) -> Result<RawTable<K, V>, CollectionAllocErr> {
         unsafe {
             let ret = RawTable::new_uninitialized_internal(capacity, fallibility)?;
-            ptr::write_bytes(ret.hashes.ptr(), 0, capacity);
+            if capacity > 0 {
```

Comment:
Instead of adding this workaround, I would prefer if here: https://github.com/rust-lang/rust/blob/9f9f2c0095cd683b94adca133f2733aa1f88bb19/src/libstd/collections/hash/table.rs#L35 

we would change `EMPTY` from a `const` to a `const fn` that just returns `NonNull::dangling().as_ptr()`. I think `EMPTY` is only used twice in the whole module.

Reply:
```rust
/// Note: when the pointer is initialized to EMPTY `.ptr()` will return
/// null and the tag functions shouldn't be used.
```

The table states this, but there are no `debug_assert!`s preventing misuse AFAICT, might be worth it to add these as well.

Reply:
I am afraid of touching `HashMap`, so I'd prefer if someone who knows *anything* about that code does those changes...

The pointer here was previously NULL, so I don't think it came from `EMPTY`.

Reply:
If capacity == 0, then `new_uninitialized_internal` can only return (https://github.com/rust-lang/rust/blob/9f9f2c0095cd683b94adca133f2733aa1f88bb19/src/libstd/collections/hash/table.rs#L688
) `hashes` with this value:

```rust
hashes: TaggedHashUintPtr::new(EMPTY as *mut HashUint),
```

which basically calls (https://github.com/rust-lang/rust/blob/9f9f2c0095cd683b94adca133f2733aa1f88bb19/src/libstd/collections/hash/table.rs#L45):

```rust
Unique::new_unchecked(EMPTY as *mut HashUint)
```

`Unique::ptr` will then just return `EMPTY as *mut HashUint` which is never null, so I have no idea how this pointer there could ever be `NULL`.

Reply:
> The table states this, but there are no debug_assert!s preventing misuse AFAICT, might be worth it to add these as well.

Well, the tag functions cannot assert anything reasonable because the last bit can be either 0 or 1.

I think this also uses values as [sentinel values](https://github.com/rust-lang/rust/pull/52508) in ways it should not.  But that really needs someone to actually dig in the code. I can open an issue about that if you want, right now I only have time to do the most obvious thing.

> I have no idea how this pointer there could ever be NULL. It should be EMPTY.

As the comment you quoted says, `ptr()` will return NULL for `EMPTY`.

If `EMPTY` was properly aligned and `ptr` did not do anything, we would not even need any test here on `write_bytes` because non-NULL aligned zero-sized writes are perfectly fine.

Reply:
>  I can open an issue about that if you want, right now I only have time to do the most obvious thing.

Yeah, this is already an improvement.

Reply:
Done, I opened https://github.com/rust-lang/rust/issues/53853.

Anything left to do in this PR?